### PR TITLE
Computing metrics per-class for imbalanced data

### PR DESCRIPTION
**Pull request by AnselmC**
# Before submitting

- ✅ Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- ✅ Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- ✅ Did you make sure to update the docs?   
- ✅  Did you write any new necessary tests?  

## What does this PR do?
Draft PR as discussed in #174.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@AnselmC how is it going here? ready for review?

**Comment by AnselmC** [comment]
@Borda Hi, so I believe it's at least ready for an initial discussion. 
One point I was struggling with/unsure about is how to handle multi-dimensional multi-class data (and their different accumulations). As I understand (please correct me if I'm mistaken), this means that a sample may belong to multiple classes encoded by a [multi-dimensional binary vector](https://pytorch-lightning.readthedocs.io/en/stable/extensions/metrics.html#input-types). Hence, when `mdmc_average="global"`, these multi-dim vectors need to be treated as a label - this seems analogous to how I've dealt with integer labels.
In the `samplewise` case, however, each sample is treated as a separate "batch" and then the metrics are computed per "batch" and averaged - I'm not sure how to correctly address this but I'm inclined to keep functionality as is for this case.

Would appreciate some input on this or pointers to more documentation for these scenarios.

File: torchmetrics/functional/classification/f_beta.py [hunk 1]
```diff
@@ -50,22 +51,32 @@ def _fbeta_compute(
     num = (1 + beta**2) * precision * recall
     denom = beta**2 * precision + recall
     denom[denom == 0.] = 1  # avoid division by 0
+    # if classes matter and a given class is not present in both the preds and the target,
+    # computing the score for this class is meaningless, thus they should be ignored
+    if average == AverageMethod.NONE and mdmc_average is None:
+        # a class is not present if there exists no TPs, no FPs, and no FNs
+        meaningless_indeces = torch.nonzero(np.logical_not(np.logical_or(tp.cpu(),
+                                                                         np.logical_or(fp.cpu(), fn.cpu()))))
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/f_beta.py:59]: please only use pytorch operations
```suggestion
        meaningless_indeces = ((tp | fn | fp) == 0).nonzero()
```

File: torchmetrics/functional/classification/f_beta.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 from typing import Optional
 
+import numpy as np
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/f_beta.py:16]:
```suggestion

```

**Review by SkafteNicki — COMMENTED** [review]
overall lgtm

**Comment by AnselmC** [comment]
@Borda @SkafteNicki should I add this for the other classification metrics as well? Or do you want separate PRs?

**Comment by Borda** [comment]
> @Borda @SkafteNicki should I add this for the other classification metrics as well? Or do you want separate PRs?

Checking the size of such change and I think of having it all here...

**Comment by SkafteNicki** [comment]
Yes, agree make all the changes here. We are talking around 6-7 metrics right?

**Comment by EduardKononov** [comment]
The snippet you provided in [the original issue](https://github.com/PyTorchLightning/metrics/issues/174#issue-859686611) 
```python
import torch
import torchmetrics

f1 = torchmetrics.F1(num_classes=2, average="none")
preds = torch.tensor([0, 0, 0])
targets = torch.tensor([0, 0, 0])
print(f1(preds, targets))
```

throws a deprecation warning (PyTorch 1.7.1):
```
/torchmetrics/functional/classification/f_beta.py:57: UserWarning: This overload of nonzero is deprecated: nonzero()
Consider using one of the following signatures instead: nonzero(*, bool as_tuple) (Triggered internally at  /opt/conda/conda-bld/pytorch_1607370172916/work/torch/csrc/utils/python_arg_parser.cpp:882.)
meaningless_indeces = ((tp | fn | fp) == 0).nonzero().cpu()
```
The line was added in 6b6996d4a4840d9a40263d4b572e03d9c9fabbe6

**Comment by AnselmC** [comment]
@Borda @SkafteNicki the affected metrics are: `accuracy`, `precision`, `recall`, `precision_recall`, `specificity`, `fbeta`, and `f1`.

File: tests/classification/test_f_beta.py [hunk 3]
```diff
@@ -129,7 +130,7 @@ def test_zero_division(metric_class, metric_fn):
     """ Test that zero_division works correctly (currently should just set to 0). """
 
     preds = torch.tensor([1, 2, 1, 1])
-    target = torch.tensor([2, 1, 2, 1])
+    target = torch.tensor([2, 0, 2, 1])
```
Comment (Borda) [inline tests/classification/test_f_beta.py:133]: why is this needed?

Comment (AnselmC) [reply tests/classification/test_f_beta.py:133]: If there is no `0` class in both the preds and the targets, the new implementation will return `nan` and not 0 as expected by this test. However, to test whether `zero_division` works correctly, all that is needed is 0 `TPs` and 0 `FPs` and >0 `FNs` (see [here](https://github.com/AnselmC/metrics/blob/22c3709c6d4e8a5d6f36c9d410f680e297548610/torchmetrics/functional/classification/f_beta.py#L44)).

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@AnselmC mind check the failing tests? :rabbit:

**Comment by AnselmC** [comment]
@Borda sorry, I was out on vacation for a week. I fixed the valid issues from deep source but not sure how to deal with `torch.tensor is not callable` (see [here](https://github.com/pytorch/pytorch/issues/24807))).
Anything else you need from me on this atm?

**Comment by Borda** [comment]
@SkafteNicki @maximsch2 mind review?

**Review by SkafteNicki — APPROVED** [review]
Great job, LGTM!
Please add entry to changelog on which metrics are affected by this enchancement :]

**Comment by AnselmC** [comment]
@SkafteNicki done!
